### PR TITLE
Refactor to remove opening browser and just return url(s)

### PR DESCRIPTION
**Pull request by nanikjava**
The WaitAndMaybeOpenService(..) method allow user to open the service
url in a browser when found, this create issue during testing as it's
opening browser and consuming resources.

The fix is to introduce a boolean flag allowing the caller to specify
whether to just print out the url or open a browser window

### CODE REVIEW

File: pkg/minikube/service/service.go [hunk 1]
```diff
@@ -267,7 +267,7 @@ func PrintServiceList(writer io.Writer, data [][]string) {
 
 // WaitAndMaybeOpenService waits for a service, and opens it when running
 func WaitAndMaybeOpenService(api libmachine.API, namespace string, service string, urlTemplate *template.Template, urlMode bool, https bool,
```
Comment (tstromberg) [inline pkg/minikube/service/service.go]: WaitAndMaybeOpenService already has too many arguments, let's not add another.

How about removing the `browser.OpenURL` call altogether, and leaving that up to `cmd/service.go` to handle. Then this function could be renamed to a more definite `WaitForService`, and be more testable.

Comment (medyagh) [reply pkg/minikube/service/service.go]: Agree with tstromberg

Comment (nanikjava) [reply pkg/minikube/service/service.go]: Cool. Will make the changes

File: pkg/minikube/service/service.go [hunk 2]
```diff
@@ -265,22 +264,24 @@ func PrintServiceList(writer io.Writer, data [][]string) {
 	table.Render()
 }
 
-// WaitAndMaybeOpenService waits for a service, and opens it when running
-func WaitAndMaybeOpenService(api libmachine.API, namespace string, service string, urlTemplate *template.Template, urlMode bool, https bool,
-	wait int, interval int) error {
+// WaitForService waits for a service, and opens it when running
```
Comment (tstromberg) [inline pkg/minikube/service/service.go]: Make comment accurate again

File: cmd/minikube/cmd/service.go [hunk 3]
```diff
@@ -68,11 +71,16 @@ var serviceCmd = &cobra.Command{
...
 			os.Exit(1)
 		}
-		err = service.WaitAndMaybeOpenService(api, namespace, svc,
+		urlString, err := service.WaitForService(api, namespace, svc,
 			serviceURLTemplate, serviceURLMode, https, wait, interval)
 		if err != nil {
 			exit.WithError("Error opening service", err)
 		}
+
+		if len(urlString) != 0 {
```
Comment (tstromberg) [inline cmd/minikube/cmd/service.go]: `WaitAndMaybeOpenService` had the following logic for opening a URL:

`if urlMode || !isHTTPSchemedURL {`

Make sure your PR preserves the same behavior as before.

File: pkg/minikube/service/service.go [hunk 4]
```diff
@@ -265,22 +264,24 @@ func PrintServiceList(writer io.Writer, data [][]string) {
...
+func WaitForService(api libmachine.API, namespace string, service string, urlTemplate *template.Template, urlMode bool, https bool,
+	wait int, interval int) (string, error) {
+
+	var urlString string
```
Comment (tstromberg) [inline pkg/minikube/service/service.go]: Just `url` is fine as a variable name. It generally isn't useful to repeat the type of an object in the object name.

See also https://github.com/golang/go/wiki/CodeReviewComments#variable-names

File: pkg/minikube/service/service_test.go [hunk 5]
```diff
@@ -903,12 +903,12 @@ func TestWaitAndMaybeOpenService(t *testing.T) {
 				servicesMap:  serviceNamespaces,
 				endpointsMap: endpointNamespaces,
 			}
-			err := WaitAndMaybeOpenService(test.api, test.namespace, test.service, defaultTemplate, test.urlMode, test.https, 1, 0)
+			_, err := WaitForService(test.api, test.namespace, test.service, defaultTemplate, test.urlMode, test.https, 1, 0)
```
Comment (tstromberg) [inline pkg/minikube/service/service_test.go]: This returns a URL now - please test the returned value of it.

File: pkg/util/utils.go [hunk 6]
```diff
@@ -200,3 +202,9 @@ func ConcatStrings(src []string, prefix string, postfix string) []string {
 	}
 	return ret
 }
+
+func OpenBrowser(urlString string) {
```
Comment (tstromberg) [inline pkg/util/utils.go]: This function doesn't add ta lot of value, so just remove it and inline the code instead. The only thing really duplicated is the error message. 

Also, we're trying to get rid of unscoped packages like the `util` package.

File: cmd/minikube/cmd/service.go [hunk 7]
```diff
@@ -68,11 +71,16 @@ var serviceCmd = &cobra.Command{
...
+			out.T(out.Celebrate, "Opening kubernetes service  {{.namespace_name}}/{{.service_name}} in default browser...", out.V{"namespace_name": namespace, "service_name": svc})
+			util.OpenBrowser(urlString)
```
Comment (tstromberg) [inline cmd/minikube/cmd/service.go]: If opening the browser fails, call `exit.WithError`

File: pkg/minikube/service/service_test.go (hunk 5 above)
Comment (nanikjava) [reply pkg/minikube/service/service_test.go]: @tstromberg Want to bounce some thoughts. In master branch the current implementation is

```
	for _, bareURLString := range serviceURL.URLs {
		urlString, isHTTPSchemedURL := OptionallyHTTPSFormattedURLString(bareURLString, https)

		if urlMode || !isHTTPSchemedURL {
			out.T(out.Empty, urlString)
		} else {
			out.T(out.Celebrate, "Opening kubernetes service  {{.namespace_name}}/{{.service_name}} in default browser...", out.V{"namespace_name": namespace, "service_name": service})
			if err := browser.OpenURL(urlString); err != nil {
				out.ErrT(out.Empty, "browser failed to open url: {{.error}}", out.V{"error": err})
			}
		}
	}

```
which means that there is possibility that there are more than 1 url, so is it correct to also
implement string array as the returned value ?

File: cmd/minikube/cmd/service.go (hunk 3 above)
Comment (nanikjava) [reply cmd/minikube/cmd/service.go]: The code that you are referring to is inside pkg/minikube/service/service.go and the only changes made was to send back the array of urls

File: pkg/minikube/service/service_test.go (hunk 5 above)
Comment (nanikjava) [reply pkg/minikube/service/service_test.go]: After going through the code and the functionality the function will have to return array of url. So have made changes to both the code and test cases.

File: pkg/minikube/service/service.go (hunk 2 above)
Comment (tstromberg) [reply pkg/minikube/service/service.go]: This method no longer opens a URL.

Comment (nanikjava) [reply pkg/minikube/service/service.go]: Comment has been updated to reflect returning url instead of opening in browser. PR title has been modified too.

**Comment by tstromberg** [comment]
Thank you for sticking through the review process!

**Comment by nanikjava** [comment]
> Thank you for sticking through the review process!

No worries. Thanks for being patience in reviewing my PRs, still learning :)

**Pull request merged**